File: zig-generator/src/prologue.zig
const std = @import("std");
const GeneratorState = @import("state.zig").GeneratorState;

pub fn emitPrologue(writer: anytype, state: *GeneratorState, num_regs: u8, sig_path: []const u8) !void {
    const buf_size: u64 = @as(u64, num_regs) * 8;
    try writer.print("// seed: 0x{X:0>16}\n\n", .{state.getSeed()});
    try writer.print(".section .data\n", .{});
    try writer.print("result_buf:  .space {d}\n", .{buf_size});
    try writer.print("result_path: .asciz \"{s}\"\n", .{sig_path});
    try writer.print(".align 6\n", .{});
    try writer.print("scratch_buf: .space 256\n\n", .{});
    try writer.print(".section .text\n", .{});
    try writer.print(".global _start\n", .{});
    try writer.print("_start:\n", .{});
    try writer.print("  ADRP x21, result_buf\n", .{});
    try writer.print("  ADD  x21, x21, :lo12:result_buf\n", .{});
    try writer.print("  ADRP x22, scratch_buf\n", .{});
    try writer.print("  ADD  x22, x22, :lo12:scratch_buf\n\n", .{});
    const scratch_id = try state.addRegion(256, GeneratorState.scratch_pointer, "scratch_buf");
    state.setScratchRegionId(scratch_id);
}

pub fn emitEpilogue(writer: anytype, num_regs: u8) !void {
    const buf_size: u64 = @as(u64, num_regs) * 8;
    try writer.print("epilogue:\n", .{});
    for (0..num_regs) |i| {
        try writer.print("  STR x{d}, [x21, #{d}]\n", .{ i, i * 8 });
    }
    try writer.print("\n    MOV  x0, #-100\n", .{});
    try writer.print("    ADRP x1, result_path\n", .{});
    try writer.print("    ADD  x1, x1, :lo12:result_path\n", .{});
    try writer.print("    MOV  x2, #577\n", .{});
    try writer.print("    MOV  x3, #0x1A4\n", .{});
    try writer.print("    MOV  x8, #56\n", .{});
    try writer.print("    SVC  #0\n\n", .{});
    try writer.print("    MOV  x1, x21\n", .{});
    try writer.print("    MOV  x2, #{d}\n", .{buf_size});
    try writer.print("    MOV  x8, #64\n", .{});
    try writer.print("    SVC  #0\n\n", .{});
    try writer.print("    MOV  x8, #93\n", .{});
    try writer.print("    MOV  x0, #0\n", .{});
    try writer.print("    SVC  #0\n", .{});
}

File: rust-generator/src/prologue.rs
use std::io::Write;
use crate::state::GeneratorState;

pub fn emit_prologue<W: Write>(out: &mut W, state: &mut GeneratorState, num_regs: u8, sig_path: &str) {
    let buf_size = num_regs as u64 * 8;
    writeln!(out, "// seed: 0x{:016X}\n", state.seed()).unwrap();
    writeln!(out, ".section .data").unwrap();
    writeln!(out, "result_buf:  .space {}", buf_size).unwrap();
    writeln!(out, "result_path: .asciz \"{}\"", sig_path).unwrap();
    writeln!(out, ".align 6").unwrap();
    writeln!(out, "scratch_buf: .space 256\n").unwrap();
    writeln!(out, ".section .text").unwrap();
    writeln!(out, ".global _start").unwrap();
    writeln!(out, "_start:").unwrap();
    writeln!(out, "  ADRP x21, result_buf").unwrap();
    writeln!(out, "  ADD  x21, x21, :lo12:result_buf").unwrap();
    writeln!(out, "  ADRP x22, scratch_buf").unwrap();
    writeln!(out, "  ADD  x22, x22, :lo12:scratch_buf\n").unwrap();
    let scratch_id = state.add_region(256, GeneratorState::SCRATCH_POINTER, "scratch_buf".to_string());
    state.set_scratch_region_id(scratch_id);
}

pub fn emit_epilogue<W: Write>(out: &mut W, num_regs: u8) {
    let buf_size = num_regs as u64 * 8;
    writeln!(out, "epilogue:").unwrap();
    for i in 0..num_regs {
        writeln!(out, "  STR x{}, [x21, #{}]", i, i as u64 * 8).unwrap();
    }
    writeln!(out, "\n    MOV  x0, #-100").unwrap();
    writeln!(out, "    ADRP x1, result_path").unwrap();
    writeln!(out, "    ADD  x1, x1, :lo12:result_path").unwrap();
    writeln!(out, "    MOV  x2, #577").unwrap();
    writeln!(out, "    MOV  x3, #0x1A4").unwrap();
    writeln!(out, "    MOV  x8, #56").unwrap();
    writeln!(out, "    SVC  #0\n").unwrap();
    writeln!(out, "    MOV  x1, x21").unwrap();
    writeln!(out, "    MOV  x2, #{}", buf_size).unwrap();
    writeln!(out, "    MOV  x8, #64").unwrap();
    writeln!(out, "    SVC  #0\n").unwrap();
    writeln!(out, "    MOV  x8, #93").unwrap();
    writeln!(out, "    MOV  x0, #0").unwrap();
    writeln!(out, "    SVC  #0").unwrap();
}
